mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -830,17 +830,558 @@ LogicalResult ThreadwiseWriteAllRewritePattern::matchAndRewrite(
+    SmallVector<StringRef, 4> nonReduceNameRefs;
+    for (auto dimAndSize : llvm::enumerate(toReduceShape)) {
+      int64_t dim = dimAndSize.index();
+      int64_t dimSize = dimAndSize.value();
+      if (dim != reduceAxis) {
+        nonReduceMergeDimSize *= dimSize;
+        nonReduceNameRefs.push_back(lowerNameRefs[dim]);
+      }
+    }
+    threadsToTensor.merge("nrDim", 0, nonReduceNameRefs);
+    threadsToTensor.passThrough({"rIter"}, {1}, {lowerNameRefs[reduceAxis]});
+    TransformMapAttr mergeTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, mergeTrMap);
+    int64_t nrThreads = (nonReduceMergeDimSize + (blockSize - 1)) / blockSize;
+    threadsToTensor.pad({"nrDim"},
+                        {0, blockSize * nrThreads - nonReduceMergeDimSize});
+    threadsToTensor.passThrough({"rIter"}, {1}, {"rIter"});
+    TransformMapAttr padTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, padTrMap);
+    threadsToTensor.unmerge({"tid", "nrIter"}, {0, 1}, "nrDim",
+                            {blockSize, nrThreads});
+    threadsToTensor.passThrough({"rIter"}, {2}, {"rIter"});
+    TransformMapAttr unmergeTrMap = threadsToTensor.get();
+
+    return rewriter.getArrayAttr({unmergeTrMap, padTrMap, mergeTrMap});
+  }
+
+  // This should only be used if product non-reduction dims is
+  // less than number threads in a block.
+  //
+  // Given a input tensor : D0, ... , Dr , ... , DN to reduce,
+  // This function creates a view that maps the space of
+  // [D0, ... , Dr , ... , DN] --> [nrtid, rtid, rIter] where
+  // nrtid = tid / product(non-reduction dims) is a reduction subgroup leader.
+  // rtid = tid % product(non-reduction dims) is thread idx within a reduction
+  // subgroup. | rtid | is the number of threads that'd participate in a
+  // reduction
+  ArrayAttr createThreadViewforNRSmallerThanThreads(
+      Location loc, ArrayRef<int64_t> toReduceShape, int64_t blockSize,
+      size_t reduceAxis, PatternRewriter &rewriter) const {
+    BottomUpTMBuilder threadsToTensor(rewriter, toReduceShape, loc);
+    SmallVector<StringRef, 4> lowerNameRefs;
+    threadsToTensor.getStartNames(lowerNameRefs);
+    int64_t nonReduceMergeDimSize = 1;
+    SmallVector<StringRef, 4> nonReduceNameRefs;
+    for (auto [dim, dimSize] : llvm::enumerate(toReduceShape)) {
+      if (dim != reduceAxis) {
+        nonReduceMergeDimSize *= dimSize;
+        nonReduceNameRefs.push_back(lowerNameRefs[dim]);
+      }
+    }
+    threadsToTensor.merge("nrDim", 0, nonReduceNameRefs);
+    threadsToTensor.passThrough({"rDim"}, {1}, {lowerNameRefs[reduceAxis]});
+    TransformMapAttr mergeTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, mergeTrMap);
+    // Distribute threads that is extra than nonReduceMergeDimSize and use them
+    // for reductions. This is a floor because max threads per non-reduction
```

Comment:
I have no idea what this comment is trying to say, please reword.

Reply:
reworded, see if it makes sense.